be/src/exec/pipeline/pipeline_driver_queue.h

```diff
@@ -52,22 +65,69 @@ class QuerySharedDriverQueue : public FactoryMethod<DriverQueue, QuerySharedDriv
         }
     }
     ~QuerySharedDriverQueue() override = default;
-    void close() override;
+    void close() override {}
 
-    static const size_t QUEUE_SIZE = 8;
-    // maybe other value for ratio.
-    static constexpr double RATIO_OF_ADJACENT_QUEUE = 1.2;
-    void put_back(const DriverRawPtr driver) override;
-    void put_back(const std::vector<DriverRawPtr>& drivers) override;
+    void put_back(const DriverRawPtr driver, bool from_dispatcher) override;
+    void put_back(const std::vector<DriverRawPtr>& drivers, bool from_dispatcher) override;
     // return nullptr if queue is closed;
-    StatusOr<DriverRawPtr> take(size_t* queue_index) override;
-    SubQuerySharedDriverQueue* get_sub_queue(size_t) override;
+    StatusOr<DriverRawPtr> take() override;
+
+    void yield_driver(const DriverRawPtr driver) override;
+
+    size_t size() override { return _size; }
 
 private:
+    static constexpr size_t QUEUE_SIZE = 8;
+    static constexpr double RATIO_OF_ADJACENT_QUEUE = 1.2;
+
     SubQuerySharedDriverQueue _queues[QUEUE_SIZE];
+
+    size_t _size = 0;
+};
+
+// DriverQueueWithWorkGroup contains two levels of queues.
+// The first level is the work group queue, and the second level is the driver queue in a work group.
+class DriverQueueWithWorkGroup : public FactoryMethod<DriverQueue, DriverQueueWithWorkGroup> {
+    friend class FactoryMethod<DriverQueue, DriverQueueWithWorkGroup>;
+
+public:
+    ~DriverQueueWithWorkGroup() override = default;
+    void close() override;
+
+    // When the driver's workgroup is not in the workgroup queue
```

Comment:
Should move the comment to interface class.

Reply:
Sounds great.